### PR TITLE
Changes to API page

### PR DESCRIPTION
**Pull request by Markus28**
I have made some changes to the API page:
- I explained the agent-environment control loop
- I replaced the CartPole example with LunarLander-v2. The reason is the following: I think that it is important to show seeding and resetting in the first minimal example. If we use CartPole, the environment will be reset every few frames and the resulting gif may look confusing. This is what it would look like: cart_pole
- I added rigorous documentation of step, reset, render in the style of readthedocs
- I removed some information on different types of spaces. I feel like this isn't part of the API and should be moved to spaces/index.md with a reference in api.md (I haven't moved anything yet)
- I added a note that `seed` is being deprecated
- I changed "Unwrapping an Environment" to "Wrappers". I think it would be neat to have a very brief section on wrappers on the API page with a link to the proper page. Though I haven't written anything about wrappers yet.

Take all of this as a mere suggestion.

### CODE REVIEW

**Comment by Markus28** [comment]
Before this pull request can be merged:
- I used a figure from OpenAI's gym documentation without reference
- I used parts of docstrings from gym without reference

Is this okay?

File: docs/pages/api/index.md [hunk 1]
```diff
@@ -12,33 +12,115 @@ env = gym.make('CartPole-v0')
...
+
+## Commonly used methods
+>### Stepping
+> `step(action: ActType) -> Tuple[ObsType, float, bool, dict]`
+> 
+>Run one timestep of the environment's dynamics. When end of episode is reached, you are responsible for calling `reset()`
+to reset this environment's state. Accepts an action and returns a tuple `(observation, reward, done, info)`.
+>
+>>**Parameters**
+>>- `action`(**object**): an action provided by the agent. This will oftentimes be a numpy array, but might also be
```
Comment (RedTachyon) [inline docs/pages/api/index.md:65]: Maybe mention that an action is often described by a `gym.Space`? Same for observations.

Actually, it might be good to mention action and observation spaces before the methods (now it's at the end), -- and that observation spaces and action spaces are described by a Space, which is typically either a Box or a Discrete (or sometimes a Dict) and so on.

File: docs/pages/api/index.md [hunk 2]
```diff
@@ -12,33 +12,115 @@ env = gym.make('CartPole-v0')
...
+>>
+>> This method returns a tuple `(observation, reward, done, info)`:
+>>- `observation` (**object**): agent's observation of the current environment.
+>> This may, for instance, be a numpy array containing the positions and velocities of certain objects.
+>>- `reward` (**float**) : amount of reward returned after previous action
+>>- `done` (**bool**): whether the episode has ended, in which case further `step()` calls will return undefined results.
+>> A done signal may be emitted for different reasons: Maybe the task underlying the environment was solved successfully, 
+>> a certain timelimit was exceeded, or the physics simulation has entered an invalid state. `info` may contain additional
+>> information regarding the reasone for a done signal.
+>>- `info` (**dict**): contains auxiliary diagnostic information (helpful for debugging, and sometimes learning)
```
Comment (RedTachyon) [inline docs/pages/api/index.md:78]: I'd also mention logging metrics

File: docs/pages/api/index.md [hunk 3]
```diff
@@ -12,33 +12,115 @@ env = gym.make('CartPole-v0')
...
 
 The output should look something like this
 
-![cartpole-no-reset](https://user-images.githubusercontent.com/28860173/129241283-70069f7c-453d-4670-a226-854203bd8a1b.gif)
+![lunar_lander](https://user-images.githubusercontent.com/15806078/153222406-af5ce6f0-4696-4a24-a683-46ad4939170c.gif)
+
+
+## Commonly used methods
+>### Stepping
+> `step(action: ActType) -> Tuple[ObsType, float, bool, dict]`
```
Comment (RedTachyon) [inline docs/pages/api/index.md:59]: For this and subsequent methods - I think I'd prefer a notation of `env.step(self, action: ActType)` or something like that, the current form looks a bit as if they were standalone functions.

File: docs/pages/api/index.md [hunk 4]
```diff
@@ -12,33 +12,115 @@ env = gym.make('CartPole-v0')
...
+>> This may, for instance, be a numpy array containing the positions and velocities of certain objects.
+>>- `reward` (**float**) : amount of reward returned after previous action
+>>- `done` (**bool**): whether the episode has ended, in which case further `step()` calls will return undefined results.
+>> A done signal may be emitted for different reasons: Maybe the task underlying the environment was solved successfully, 
+>> a certain timelimit was exceeded, or the physics simulation has entered an invalid state. `info` may contain additional
+>> information regarding the reasone for a done signal.
+>>- `info` (**dict**): contains auxiliary diagnostic information (helpful for debugging, and sometimes learning)
+
+>### Resetting
+>`reset(*, seed: Optional[int] = None, return_info: bool = False, options: Optional[dict] = None,) -> Union[ObsType, tuple[ObsType, dict]]`
```
Comment (RedTachyon) [inline docs/pages/api/index.md:81]: Remove the trailing comma in the argument list. 

The return type would be better as `ObsType | tuple[ObsType, dict]` imo (and fortunately in docs we don't need to worry about the python version)

File: docs/pages/api/index.md [hunk 5]
```diff
@@ -12,33 +12,115 @@ env = gym.make('CartPole-v0')
...
+>Resets the environment to an initial state and returns an initial observation.
+>>**Parameters**
+>>- `seed`(**int** or **None**): The seed that is used to initialize the environments PRNG. If the environment does not already
+>> have a PRNG and `seed=None`(the default option) is passed, a seed will be chosen from some source of entropy (e.g. timestamp 
+or /dev/urandom). However, if the environment already has a PRNG and `seed=None` is pased, the PRNG will *not* be reset. 
+If you pass an integer, the PRNG will be reset even if it already exists. Usually, you want to pass an integer *right after the environment
+has been initialized and then never again*. Please refer to the minimal example above to see this paradigm in action.
+>>- `return_info`(**bool**): If true, return additional information along with initial observation. This info should be analogous
+to the info returned in `step`
+>>- `options`(**dict** or **None**):
```
Comment (RedTachyon) [inline docs/pages/api/index.md:92]: [...]: Additional information to specify how the environment is reset (optional, depending on the specific environment)

(feel free to rephrase this)

File: docs/pages/api/index.md [hunk 6]
```diff
@@ -12,33 +12,115 @@ env = gym.make('CartPole-v0')
...
+>> a certain timelimit was exceeded, or the physics simulation has entered an invalid state. `info` may contain additional
+>> information regarding the reasone for a done signal.
+>>- `info` (**dict**): contains auxiliary diagnostic information (helpful for debugging, and sometimes learning)
+
+>### Resetting
+>`reset(*, seed: Optional[int] = None, return_info: bool = False, options: Optional[dict] = None,) -> Union[ObsType, tuple[ObsType, dict]]`
+>
+>Resets the environment to an initial state and returns an initial observation.
+>>**Parameters**
+>>- `seed`(**int** or **None**): The seed that is used to initialize the environments PRNG. If the environment does not already
```
Comment (RedTachyon) [inline docs/pages/api/index.md:85]: environment's*

File: docs/pages/api/index.md [hunk 7]
```diff
@@ -12,33 +12,115 @@ env = gym.make('CartPole-v0')
...
+or /dev/urandom). However, if the environment already has a PRNG and `seed=None` is pased, the PRNG will *not* be reset. 
+If you pass an integer, the PRNG will be reset even if it already exists. Usually, you want to pass an integer *right after the environment
+has been initialized and then never again*. Please refer to the minimal example above to see this paradigm in action.
+>>- `return_info`(**bool**): If true, return additional information along with initial observation. This info should be analogous
+to the info returned in `step`
+>>- `options`(**dict** or **None**):
+>
+>>**Returns**
+>>
+>>This method will return an observation of the initial environment state. If `return_info=True` is passed, the method returns a tuple
```
Comment (RedTachyon) [inline docs/pages/api/index.md:96]: Imo add a few lines in the same format as in `env.step`, i.e.
`obervation` (**object**): blahblah
`info` (**dict**): (only if `return_info=True`) blahblah

File: docs/pages/api/index.md (hunk 4 above)
Comment (RedTachyon) [reply docs/pages/api/index.md:81]: Also applicable to later usages of `Union`

File: docs/pages/api/index.md [hunk 8]
```diff
@@ -12,33 +12,115 @@ env = gym.make('CartPole-v0')
...
+>> `(observation, info)`, where `observation` is the actual observation and `info` is some auxiliary information complementing `observation`.
+>> The dictionary `info` should be analogous to the `info` returned by `step`. If `return_info=False` is passed, only `observation` is returned.
+
+>### Rendering
+> 
+>`render(mode: str="human") -> Optional[Union[numpy.ndarray, str]]`
```
Comment (RedTachyon) [inline docs/pages/api/index.md:102]: np.ndarray

File: docs/pages/api/index.md [hunk 9]
```diff
@@ -12,33 +12,115 @@ env = gym.make('CartPole-v0')
...
+>>**Parameters**
+>>- `mode`(**str**): This parameter specifies how the environment is rendered. The set of supported modes varies per environment. (And some
+environments do not support rendering at all.) By convention, if mode is:
```
Comment (RedTachyon) [inline docs/pages/api/index.md:105]: [...] some third-party environments **may** not support [...]

I'm not 100% certain here, but a part of the API specification is that there should be some way of rendering stuff. Of course it's not always the case, but let's not encourage it. Also, all built-in envs do support rendering

File: docs/pages/api/index.md [hunk 10]
```diff
@@ -12,33 +12,115 @@ env = gym.make('CartPole-v0')
...
+>>      - "human": render to the current display or terminal and return nothing. Usually for human consumption.
+>>      - "rgb_array": Return an numpy.ndarray with shape (x, y, 3), representing RGB values for an x-by-y pixel image, suitable
```
Comment (RedTachyon) [inline docs/pages/api/index.md:107]: np.ndarray

And maybe change `x, y` to `width, height`. But check which one is which. I'm frankly not sure.

Comment (RedTachyon) [reply docs/pages/api/index.md:107]: (this will also remove the need for the `x-by-y pixel image` part

File: docs/pages/api/index.md [hunk 11]
```diff
@@ -61,51 +143,13 @@ print(env.observation_space.high)
 print(env.observation_space.low)
 #> array([-4.8000002e+00, -3.4028235e+38, -4.1887903e-01, -3.4028235e+38], dtype=float32)
 ```
-- There are multiple `Space` types available in Gym:
```
Comment (RedTachyon) [inline docs/pages/api/index.md:64]: Why is this removed?

File: docs/pages/api/index.md [hunk 12]
```diff
@@ -12,33 +12,115 @@ env = gym.make('CartPole-v0')
...
 env.close()
 ```
 
 The output should look something like this
 
-![cartpole-no-reset](https://user-images.githubusercontent.com/28860173/129241283-70069f7c-453d-4670-a226-854203bd8a1b.gif)
+![lunar_lander](https://user-images.githubusercontent.com/15806078/153222406-af5ce6f0-4696-4a24-a683-46ad4939170c.gif)
+
+
+## Commonly used methods
```
Comment (RedTachyon) [inline docs/pages/api/index.md:57]: I'd change this to maybe "Standard methods" or "Common API methods" or something like that

Basically we want to imply that this is a complete list of methods shared between **all** gym environments, not just that they happen to be used a lot

**Review by RedTachyon — CHANGES_REQUESTED** [review]
Overally looks pretty good! Still needs a bunch of modifications, but I think it goes in a good direction.

A few comments that aren't in the diff:
- Maybe a few more words about wrappers, what they can do etc
- How confident are we in the `Playing within an environment` section? The whole `play.py` section seems unmaintained since 2019. I'd remove this section for now, unless someone is actually confident that it works

Note: I don't have a way to quickly visualize it, so I'm just gonna assume that it renders the way we expect it to render. A separate testing URL would be useful, but afaik we don't have that?

File: docs/pages/api/index.md (hunk 9 above)
Comment (Markus28) [reply docs/pages/api/index.md:105]: That's entirely possible. This was copied from a docstring so we should probably make a corresponding change to that docstring

File: docs/pages/api/index.md (hunk 2 above)
Comment (Markus28) [reply docs/pages/api/index.md:78]: Also copied from a docstring. I will make a corresponding change to that docstring. However, I'm not completely sure what you mean by metrics. We are not talking about losses, right? Could you give an example?

**Comment by Markus28** [comment]
About play: This was added in a different pull request (not by me). I tried it out myself and it seems (superficially) to work for Atari. Idk whether it should be left in there but imo it should be mentioned somewhere. 
I only learned about it from the pull request and I think it's a really neat feature that could have saved me some trouble in the past, had I known it.

File: docs/pages/api/index.md (hunk 1 above)
Comment (Markus28) [reply docs/pages/api/index.md:69]: I am now briefly mentioning `observation_space` and `action_space` after the minimal example. They are then discussed in more detail in the section "Additional Environment API", followed by a section on Spaces. Do you think that's sufficient?
I didn't want to get lost in all the intricate details on spaces at the very beginning of the page so I pushed some of it further to the back.

File: docs/pages/api/index.md (hunk 2 above)
Comment (RedTachyon) [reply docs/pages/api/index.md]: As an example from the top of my mind - I'm training some agents to move from point A to point B, so in the info dict I pass the current distance to the destination; then I log that information and use it to track the training progress, since it's a measure of the agent's performance (i.e. a metric)

**Comment by RedTachyon** [comment]
There's one last comment about the description of `info`, other than that looks good

**Comment by Markus28** [comment]
Thanks @RedTachyon. Did you mean this comment: https://github.com/Farama-Foundation/gym-docs/pull/77#discussion_r806634748? I am now mentioning logging and I have added a few examples of stuff that might be contained in `info`. I hope that clarifies things.

**Comment by Markus28** [comment]
One thing before this can be merged: Is this fine https://github.com/Farama-Foundation/gym-docs/pull/77#issuecomment-1036244376?

**Comment by RedTachyon** [comment]
Yep, that was it, I think it can be merged

**Comment by RedTachyon** [comment]
And regarding the copyrights (?) thing, I'm very much not a lawyer, but my intuition is that it's fine since the website is explicitly tied to gym, and the two figures are (1) a bunch of lines and words, and (2) a rendering of the env.

**Comment by trigaten** [comment]
Please resolve merge conflicts.

**Comment by Markus28** [comment]
@trigaten the conflict is resolved. Idk where it came from.

**Comment by trigaten** [comment]
@Markus28 Just to be sure, is this all finished and ready to be merged?
